Add Video Popover

DiemenDesign opened the pull request:
This PR, adds VideoPopover.js and events to allow Videos to have the ability to have a popover for adding editing options. So far only a Remove button is added.

When inserting a video a div with a class of .note-video-clip-wrapper is wrapped on the iframe video element with an invisible after element to allow clicking on the video (we don't usually play videos when editing content). The image highlighter/resizer appears over the video, can be resized, but it doesn't affect the video only it's own element at this stage unless someone else wants to add this in for resizing videos.

DiemenDesign commented:
@hackerwins there's a strange error that travis stops at in the VideoDialog.js where the script builds the iframe for the v.qq.com service:
PhantomJS 2.1.1 (Linux 0.0.0) bs:module.VideoDialog #createVideoNode should execute when insert v.qq.com FAILED
	expected 'DIV' to equal 'IFRAME'

I'm not sure how to fix it.

hackerwins commented:
@DiemenDesign I think the error was related to the changes below.
https://github.com/summernote/summernote/pull/2666/files#diff-6567c42e304af4f81b874d4cdedf302fR140.

DiemenDesign commented:
That links shows all the changes I made to add the popover.
The error that Travis now shows is in the VideoDialog.js file, and something to do with the v.qq.com service when the script builds the iframe. I thought it might have been the unintentional `    $video.attr('frameborder', 0);` before the const that builds the wrapper I'd left in, but I removed it, and the error still persists.

DiemenDesign commented:
@hackerwins I can't work out this issue, I'm going to need help to get this to pass the tests.

hackerwins commented:
@DiemenDesign Thank you for your efforts. I'll check the error this week end.

DiemenDesign commented:
@hackerwins Thank you, I'll be interested to know what causes it for future reference.
If you look at some of the other edits, You'll see I even tried moving the assembly of the wrapper element to the outside of that function with no luck.

hackerwins commented:
Before this PR, Video was created with `<iframe>` tag. But below line wrap video tag with `div.note-video-clip-wrapper`. So the test that was merged by #2048 is broken. That means spec is changed by this PR.

https://github.com/summernote/summernote/pull/2666/files#diff-6567c42e304af4f81b874d4cdedf302fR137

So I hope you think about why the video wrapped by div. If we can do same thing without the note-video-clip-wrapper, I think we should remove it.

DiemenDesign commented:
The reason it's wrapped with the div is to allow the selection styling to work. Without it, videos have a z-index within their iframe that places them above all content. I tried so many different ways to not have the wrapper div, only to find out that it's needed due to the z-index.

DiemenDesign commented:
I would prefer to do it without the extra markup if I could.

DiemenDesign commented:
Shame too, as npm compiles just fine, it's just Travis reporting the error.

I'll more of a play with not using extra markup and see if I can get something to work here.

hackerwins commented:
For your information, You can run tests with `npm run test` on your local pc.
https://github.com/summernote/summernote/blob/develop/CONTRIBUTING.md#test-summernote

It's a good for saving time to run the test on local pc before pushing the code.

DiemenDesign commented:
Ah ok, I've just been building it locally. :+1:

DiemenDesign commented:
Ok, I've done some research, and it turns out that it's how the Browser loads the elements into the DOM. Because the videos not matter how they are embedded end up loading last in the order of elements, they get added last giving them the highest z-index order. I have however worked out a bit of a cheat way to get the popover to work, by having a :hover on the .note-video-clip class, and forcing the border to grow in size, thus shrinking the video and exposing a clickable area. I'm just styling it up so it looks decent, and has an :after element text to let people know it's clickable.

DiemenDesign commented:
Just some notes on the lastest updates:

Added styling that gets appended to the iframe to allow hovering which alters the left side padding allowing that area to be clicked on to allow the videoPopover to open.

I tried to add the ability when a video is highlighted and the insertVideo toolbar button is clicked to open the insertVideoDialog that it contains the URL of the video allowing it to be changed. I was able to grab the URL, but couldn't work out how to change that, so for now I will leave that to the videoAttributes plugin which I need to modify to work correctly and add the extra video services.

I added some form styling for the Lite version mainly for formatting form element inside dialogs so they look nicer, and format properly, and to prepare for integrating the imageAttributes as per @lqez's request a while ago. The styling follows the Bootstrap styling but with a prefix of .note where applicable.

Once this PR is approved and Merged, I will go through the Dialog's, mainly for the Lite version and fix their formatting and add any necessary ARIA attributes for screen readers.

hackerwins commented:
@DiemenDesign 
Thank you for your researching. I'll go through some tests.

hackerwins commented:
I tested video popover and found that click and hover events do not work with iframes in latest Chrome and Firefox. Events such as clicking the play button in the iframe occur instead of the event attached in the editor.

DiemenDesign commented:
Welcome to my dilema. Trouble is wrapping the iFrame with any other element causes the error when building the video elements.
I did have a look at how other editors handle video's, and the wrap the iframe which as you've discovered allows the parent element to be rendered above the iframe.
Mind you, when I build the editor locally, it all works with the wrapper, it's only the Test that throws the error. And I checked in Chrome Dev Tools, and Firefox Qauntum, and there's no errors thrown when using a wrapper.

DiemenDesign commented:
Oh, I think a wrapper would look much nicer rather than what I've currently implemented.

hackerwins commented:
Yes. I agreed with your opinion. And we need to do something that div does not work like paragraph.

hackerwins commented:
Thank you for your interest in core functionality of Summernote.

lqez commented:
Hi @DiemenDesign, how could we handle this again? Please let me know the entry point to keep up.

DiemenDesign commented:
With the significant changes to the core since this PR was originally submitted, it might be wise to revisit all the changes this PR does. I've been a little busy lately. If someone wants to take over the changes and fix them for the current version of Summernote I would be cool with that. For me it's about adding the functionality for all the users, not just myself.
In fact, I should revisit doing the ARIA stuff, as I've noticed there's stuff missing, and stuff that's being added that is either wrong, or not needed.

lqez commented:
I'm closing this PR. I hope we can handle this feature in the future again.
